File: leetcode_ly/java/List/ReverseLinkedListTwo.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        //left->ptr1->...->right->ptr2
        if(head == null || m == n) return head;
        ListNode left = head;
        ListNode right = head;

        if(m == 1) left = null;
        while(m > 2){
            left = left.next;
            m--;
        }
        //获取要翻转的头结点
        ListNode ptr1 = head;
        if(left != null) ptr1 = left.next;
        while(n != 1){
            right = right.next;
            n--;
        }
        ListNode ptr2 = right.next;
        
        ListNode p = ptr2;
        ListNode q = ptr1;
        while(ptr1 != right){
            ptr1 = ptr1.next;
            q.next = p;
            p = q;
            q = ptr1;
        }
        ptr1.next = p;
        if(left == null) return ptr1;
        else{
            left.next = right;
            return head;
        }
    }
}

File: leetcode_ly/c++/List/ReverseLinkedListTwo.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if(!head || !head->next || m == n) return head;
        ListNode* front = new ListNode(0);
        front->next = head;
        ListNode* low = front;
        if(m != 1){
            for(int i = 0; i < m - 1; i++){
                low = low->next;
            }
        }
        ListNode* q = low->next;
        ListNode* node = q;
        low->next = NULL;
        ListNode* p = NULL;
        ListNode* r = q->next;
        int cnt = 1;
        while(r && cnt <= n - m){
            q->next = p;
            p = q;
            q = r;
            r = r->next;
            cnt++;
        }
        q->next = p;
        low->next = q;
        node->next = r;
        return front->next;
    }
};